messaging/src/main/java/org/axonframework/messaging/responsetypes/ResponseType.java

```diff
@@ -38,20 +38,43 @@
  */
 public interface ResponseType<R> extends Serializable {
 
+    int NO_MATCH = 0;
+    int MATCHES_SINGLE = 1;
+    int MATCHES_LIST = 1024;
+
     /**
      * Match the query handler its response {@link java.lang.reflect.Type} with the {@link ResponseType} implementation
-     * its expected response type {@code R}. Will return true if a response can be converted based on the given
-     * {@code responseType} and false if it cannot.
+     * its expected response type {@code R}. Will return true if a response can be converted based on the given {@code
+     * responseType} and false if it cannot.
      *
      * @param responseType the response {@link java.lang.reflect.Type} of the query handler which is matched against
      * @return true if a response can be converted based on the given {@code responseType} and false if it cannot
      */
     boolean matches(Type responseType);
 
     /**
-     * Converts the given {@code response} of type {@link java.lang.Object} into the type {@code R} of this
-     * {@link ResponseType} instance. Should only be called if {@link ResponseType#matches(Type)} returns true.
-     * It is unspecified what this function does if the {@link ResponseType#matches(Type)} returned false.
+     * Defines the match and its priority. The greater the value above 0, the better the match. This is particularly
+     * useful for {@link MultipleInstancesResponseType MultipleInstancesResponseTypes} when there are match on a both
+     * multiple and single instance types. Lists should be given priority for handling.
+     * <p>
+     * See also the constants defined to indicate match priority: {@link ResponseType#MATCHES_LIST}, {@link
+     * ResponseType#MATCHES_SINGLE} and {@link ResponseType#NO_MATCH}.
+     *
+     * @param responseType the response {@link java.lang.reflect.Type} of the query handler which is matched against
+     * @return {@link ResponseType#NO_MATCH} if there is no match, greater than 0 if there is a match. Highest match
+     * should win when choosing a query handler.
+     */
+    default Integer matchPriority(Type responseType) {
+        if(matches(responseType)) {
```

Comment:
Nit: Indentation is slightly off here.
```suggestion
        if (matches(responseType)) {
```